### DISCUSSION TITLE
Agent interaction and maintaining lists of agents

### DISCUSSION THREAD

**Discussion by nataliedavis** [JoelForamitti/agentpy#31] [Q&A] [answered]
Hi,

I am trying to teach myself AgentPy by rebuilding the classic Wolf-Sheep predation model. So far it is going pretty well, but I had a couple questions about how I had implemented agent storage in the landscape, and agent interaction. My code is below (I removed some of the lines for brevity but all relevant code is shown), and my questions are:

1. Is there a cleaner way than what I did for an agent to identify other agents of another type (so not technically 'neighbors' since that seemed to only return agents of the same type), besides using a 'species' or similar variable and a filter?
2. Is there a better way to store agents within a spatial landscape, so they can access each other (related to question 1), and updates to/removals of agents by other agents are reflected in the lists?

```
def setup(self):
        self.landscape = ap.Grid(self, (self.p.size, self.p.size), torus=True)
        self.wolves = ap.AgentDList(self, self.p.wolf_popsize, Wolf, self.landscape)
        self.landscape.add_agents(self.wolves, random=True)

def step(self):
        current_sheep = self.landscape.agents.to_list().select(self.landscape.agents.species == "sheep")
        current_wolves = self.landscape.agents.to_list().select(self.landscape.agents.species == "wolf")
        if len(current_sheep) > 0:
           # Sheep do stuff
        if len(current_wolves) > 0:
            current_wolves.eat_sheep()
        self.grass.regrow()
        if len(current_wolves) == 0 and len(current_sheep) == 0:
            self.stop()
```

In the method for wolves eating sheep, I have:

```
def eat_sheep(self):
        # Identify sheep on same grid patch
        poss_prey = self.environment.neighbors(self, distance=0).to_list()
        if len(poss_prey.select(poss_prey.species == "sheep")) > 0:
            # If there are any sheep, grab one
            prey = poss_prey.select(poss_prey.species == "sheep").random().to_list()[0]
            # Remove that sheep from the list of sheep in the grid
            self.environment.remove_agents(prey)
            # Update energy reserves
            self.energy += self.gain_from_food
```

Thanks in advance for any advice!

**Comment by jofmi** [comment, marked as answer]
Hi, sorry for the late answer!

I would try to create two seperate landscapes `landscape_sheep` and `landscape_wolfs`.
Having a grid with mixed agents is only useful if you want to query mixed lists.

Lists like `current_sheep` would then simply be `landscape_sheep.agents`.
In `eat_sheep()` you can then identify sheep on the same position as follows:

```
pos = landscape_wolfs.position[self]
poss_prey = landscape_sheep.agents[pos]
if len(poss_prey) > 0:
    prey, = poss_prey.random()
    landscape_sheep.remove_agents(prey)
```

(The comma in `prey, = poss_prey.random()` achieves the same as `.to_list()[0]`)

If you are interested, we could add the finished model to the agentpy model library!

**Reply by nataliedavis** [reply]
No worries, thanks so much for getting back to me! That makes sense. I will give it a try and let you know if I have any further questions.
